src/graph.rs

```diff
@@ -460,6 +467,153 @@ where
 
         self.index = Some(index);
     }
+
+    // creates an AGraph from a list of 'good' addresses and
+    // the Edges array.
+    // For each edge, if both nodes are in the good list,
+    // a connection is created for each node.
+    // The AGraph has a equivalent information as the edges,
+    // but is just a node-centric view (required for closeness)
+    pub fn create_agraph(&self, addresses: &Vec<T>) -> AGraph {
+        let num_nodes = addresses.len();
+        let mut agraph: AGraph = AGraph::new();
+        for _ in 0..num_nodes {
+            agraph.push(Vertex::new());
+        }
+
+        // For all our edges, check if the nodes are in our address list
+        // We use the value of the addresses to find the index
+        // From then on, it's all integer indices for us
+        for edge in self.edges.iter() {
+            let source = *edge.source();
+            let target = *edge.target();
+
+            let source_result = addresses.iter().position(|&r| r == source);
+            if source_result.is_none() {
+                continue;
+            }
+
+            let target_result = addresses.iter().position(|&r| r == target);
+            if target_result.is_none() {
+                continue;
+            }
+
+            let source_index = source_result.unwrap();
+            let target_index = target_result.unwrap();
+            agraph[source_index].push(target_index);
+            agraph[target_index].push(source_index);
+        }
+        agraph
+    }
+
+    // This method returns the closeness and betweenness for a given AGraph.
+    //
+    // Closeness:
+    //   for each node
+    //     for all other nodes
+    //       find all shortest paths to that other node
+    //         accumlate all path lens
+    //         accumulate number of paths.
+    //   compute average path length
+    //
+    //
+    //  Betweenness:
+    //    When a shortest path is found
+    //      for all nodes in-between (i.e., not an end point)
+    //        increment their betweenness value
+    //
+    pub fn compute_betweenness_and_closeness(&self, agraph: &AGraph) -> (Vec<u32>, Vec<f64>) {
+        let num_nodes = agraph.len();
+
+        let mut betweenness: Vec<u32> = vec![0; num_nodes];
+        let mut closeness: Vec<f64> = vec![0.0; num_nodes];
+        let mut total_path_length: Vec<u32> = vec![0; num_nodes];
+        let mut num_paths: Vec<u32> = vec![0; num_nodes];
+
+        for i in 0..num_nodes - 1 {
+            let mut visited: Vec<bool> = vec![false; num_nodes];
+            let mut found_or_not: Vec<bool> = vec![false; num_nodes];
+            let mut search_list: Vec<usize> = Vec::new();
+
+            // mark node i and all those before i as visited
+            for j in 0..i + 1 {
```

Comment:
Can also be written, not sure if it's clearer. 
```suggestion
            for j in 0..=i {
```